Modbus bit_switch and bit_sensor implementation

yury-sannikov opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
This PR adds new functionality and should not introduce any breaking changes.
:exclamation: This PR is based on the @janiversen PR https://github.com/home-assistant/core/pull/46591

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
`bit_switch` allows mapping Modbus holding register bits to a HomeAssistant Switch. All switch operations such as on, off, and toggle works against individual bits of a Modbus holding register. You may use the Modbus input register as well, however, a switch will remain read-only.

`bit_sensor` is pretty similar to the `bit_switch`. However, with the `bit_sensor` you are not limited to 16bit Modbus register length. You may specify any arbitrary count and specify the bit number of the interest.

Since multiple sensors or switches can read from the same Modbus register, a `functools.lru_cache` caching was added to cache Modbus register for one second. This should be enough to avoid re-reads. Any write operation to any of the registers reset the cache.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
modbus:
  - name: powermain
    type: tcp
    host: 192.168.5.222
    port: 2020
    sensors:
      - name: pwrmain_input_status
        slave: 16
        address: 512
        input_type: holding
    bit_switches:
      - name: pwrmain_shut_off_secondary_ups_load
        slave: 16
        address: 516
        command_bit_number: 0
      - name: pwrmain_airconditioneer_feed_from_solar_invertor
        slave: 16
        address: 516
        command_bit_number: 1
    bit_sensors:
      - name: pwrmain_main_feed_under_over_voltage
        slave: 16
        address: 514
        bit_number: 1
      - name: pwrmain_controller_fault_watchdog_took_over
        slave: 16
        address: 514
        bit_number: 2
      - name: pwrmain_main_feed_phase_order_error
        slave: 16
        address: 514
        bit_number: 6
      - name: pwrmain_watchdog_failure
        slave: 16
        address: 514
        bit_number: 7
      - name: pwrmain_emergency_feed_outage
        slave: 16
        address: 514
        bit_number: 10
      - name: pwrmain_reserve_feed_outage
        slave: 16
        address: 514
        bit_number: 11
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
https://github.com/home-assistant/home-assistant.io/pull/16831

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

yury-sannikov committed:
implemend bit_switch and bit_sensor

yury-sannikov commented:
cc @nagyrobi

janiversen commented:
Look good from a first view, I will go into detail a bit later

yury-sannikov commented:
Cool, then I will start working on the documentation PR. lmk if you feel we can use better names

janiversen commented:
Remark I have an open PR for documentation as well, yours should extend my PR.

yury-sannikov commented:
Documentation PR has been added based on the https://github.com/home-assistant/home-assistant.io/pull/16678 PR

janiversen commented:
Hi I wanted to do a proper review, but for some reason I cannot exclude my PR so I review only your changes. As soon as my PR is merged (just waiting for a maintainer to do it), then I swiftly review your PR.

I have downloaded it too, and it works on my Mac.

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -39,6 +49,14 @@
 
 
+@lru_cache(maxsize=32)
+def _read_cached(client, what, unit, address, count, ttl_bucket):
```
I really do not see the need for this added complexity, it does not seem to make your bit implementation easier, and it complicates the rest. I suggest to remove the lru_cache part (unless you have some good arguments, why not).

yury-sannikov replied:
If I have a configuration where each sensor/switch has the same address (in the case below it's 514) I have N read calls of the same Modbus register to the Modbus Client. If I have 16 sensors attached to the same address, the integration will end up with 16 reads of the same value through the Modbus.
This read cache holds the first read value for one second. This helps avoid making sequential calls. We can remove the cache, however, then we will be making the same read calls multiple times. This is not a big deal, however, may slow down updates.
```
    bit_sensors:
      - name: pwrmain_main_feed_under_over_voltage
        slave: 16
        address: 514
        bit_number: 1
      - name: pwrmain_controller_fault_watchdog_took_over
        slave: 16
        address: 514
        bit_number: 2
      - name: pwrmain_main_feed_phase_order_error
        slave: 16
        address: 514
        bit_number: 6
```

yury-sannikov replied:
Good idea. Let me think about how to do that. I will revert `_read_cached` changes

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -189,41 +212,56 @@ def connect(self):
         """Read coils."""
         with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_coils(address, count, **kwargs)
```
I prefer the old way, unless have some good arguments as to why not.

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -189,41 +212,56 @@ def connect(self):
     def read_discrete_inputs(self, unit, address, count):
         """Read discrete inputs."""
         with self._lock:
-            kwargs = {"unit": unit} if unit else {}
```
again

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -189,41 +212,56 @@ def connect(self):
             kwargs = {"unit": unit} if unit else {}
             self._client.write_coil(address, value, **kwargs)
+            # invalidate cache on write
+            _read_cached.cache_clear()
```
why does a write invalidate the cache ?? in many setup the inputs are totally independent from the outputs.

Looking at this I think caching will cause a lot more problems. when read_* is called we want a modbus transaction to take place.

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -189,41 +212,56 @@ def connect(self):
             kwargs = {"unit": unit} if unit else {}
             self._client.write_register(address, value, **kwargs)
+            # invalidate cache on write
+            _read_cached.cache_clear()
```
again

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -189,41 +212,56 @@ def connect(self):
             kwargs = {"unit": unit} if unit else {}
             self._client.write_registers(address, values, **kwargs)
+            # invalidate cache on write
+            _read_cached.cache_clear()
```
again.

In general I think all your changes to this file apart from the 2 loops either needs a lot more explanation and test, or be removed.

janiversen commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -56,9 +74,14 @@ def modbus_setup(
             (CONF_COVER, CONF_COVERS),
+            (CONF_BINARY_SENSOR, CONF_BINARY_SENSORS),
+            (CONF_SENSOR, CONF_SENSORS),
+            (CONF_SENSOR, CONF_BIT_SENSORS),
```
traditionally every platform have its own domain and own file, but that is not a strong demand it depends a lot more on the added complexity.

yury-sannikov replied:
Lmk your best judgment on that, I can split it, though the file size, seems, manageable.
Also, I assume I can't make (and it make no sense in this case) a new domain for, let's say, a bit_sensor.

yury-sannikov replied:
Oh, cool, for some reason I thought CONF_BINARY_SENSOR is a built-in constant.

yury-sannikov replied:
Hi @janiversen, maybe you can help me with the issue I can't find any documentation about:
I added the following into the `const.py`
```
# bit_switch.py
CONF_BIT_SWITCH = "bit_switch"
CONF_BIT_SWITCHES = "bit_switches"
```
then in a `modbus.py` I have: 
```
        # load platforms
        for component, conf_key in (
            ....
            (CONF_BIT_SWITCH, CONF_BIT_SWITCHES),
        ):
            if conf_key in conf_hub:
                load_platform(hass, component, DOMAIN, conf_hub, conf_hub)
```
Also, I have `bit_switch.py` which exposes `DOMAIN` variable eq to `modbus` and has `async_setup_platform` function in it. Pretty much the same set up as `switch.py` has, except file name is `bit_switch`
When I try to run a test: 
```
    BIT_SWITCH_DOMAIN = "bit_switch"
    switch_name = "modbus_test_bit_switch"
    state = await base_test(
        hass,
        {
            CONF_NAME: switch_name,
            CONF_INPUT_TYPE: CALL_TYPE_REGISTER_INPUT,
            CONF_ADDRESS: 1234,
            CONF_SLAVE: 1,
            CONF_COMMAND_BIT_NUMBER: 5,
        },
        switch_name,
        BIT_SWITCH_DOMAIN,
        CONF_BIT_SWITCHES,
        CONF_REGISTERS,
        regs,
        expected,
        method_discovery=True,
        scan_interval=5,
    )
    assert state == expected
```
I'm getting 
`ERROR    homeassistant.setup:setup.py:138 Setup failed for bit_switch: Integration not found.`

What I'm missing? I assume I have to register `bit_switch` somehow. Your help would be very appreciated.

yury-sannikov replied:
Extracting bit sensor/switch into separate files will require extracting the base switch/sensor to the separate file as well to avoid the import cycle. That way we will have a bunch of short files. If you prefer that way, LMK, I will extract `base_swith` `base_sensor` `register_sensor` `bit_sensor` `register_swith` `bit_switch` `coil_swith` and leave schemas and `async_setup_platform` in the `sensor` and `switch`

janiversen replied:
No it will not !! see how we use base switch today. The idea is to keep functionality that belong together in the same file. Therefore sensor and base_sensor belong in one file, whereas specializations of sensor belongs in other files. Same goes for switch and base_switch. 

Cover and Climate are good examples of how to use a base class.

Do NOT touch the schemas in switch/sensor etc. those are the old configuration style and frozen. All new configuration changes goes into __init__.py

So all in all do not touch sensor.py or switch.py that should not be needed to implement bit_sensor.

The reasons for this grouping are
- coverage tell more precisely where tests are missing
- Any problem in bit_sensor that are not in other sensors are to be solved in that file.

janiversen replied:
If you look from the HA side as opposed to the device side a bit_sensor is actually a binary_sensor and a bit_switch is just a switch. So the domains are BINARY_SENSOR and SWITCH.

janiversen reviewed (changes requested):
Your changes in sensor/switch needs some more thinking. I like the idea of a base class.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -58,22 +61,23 @@ async def base_test(
+    read_result = (
+        ReadResult(register_words)
+        if register_words
+        else ModbusException("Modbus error")
```
you are testing the test setup ?

janiversen replied:
Ahhh that is a good idea, I like that.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -118,8 +122,24 @@ async def base_test(
+            args, kvargs = kvargs["write_register"]
+            mock_sync.write_register.assert_called_once_with(*args, **kvargs)
+
         return hass.states.get(entity_id).state
```
In case of "call_service" what does state contain.

janiversen replied:
Also it seems added complexity to test both read and write in the same function.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -118,8 +122,24 @@ async def base_test(
         entity_id = f"{entity_domain}.{device_name}"
+
+        # Call an arbitrary service
+        if "call_service" in kvargs:
```
Looks like you are allowing both a read and a write test in the same call, the idea of this function is one call pr test.

yury-sannikov replied:
Yeah, this test design might be confusing. Let me rethink it based on the cache decision

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -118,8 +122,24 @@ async def base_test(
+        # check write_register call
+        if "write_register" in kvargs:
+            args, kvargs = kvargs["write_register"]
+            mock_sync.write_register.assert_called_once_with(*args, **kvargs)
```
We should check that the values sent to pymodbus (bytes) correspond the to the parameters in the original call (which could be e.g. STATE_ON).

janiversen reviewed (changes requested):
Second part of review. I have downloaded your PR, to isolate your changes, making it a bit harder to review, but I wanted to give you some feedback.

yury-sannikov commented on tests/components/modbus/conftest.py:
```diff
@@ -44,6 +46,7 @@ async def base_test(
     check_config_only=False,
     config_modbus=None,
     scan_interval=None,
+    **kvargs,
```
Ok, make sense

yury-sannikov commented on tests/components/modbus/conftest.py:
```diff
@@ -58,22 +61,23 @@ async def base_test(
     ):
 
-        # Setup inputs for the sensor
-        read_result = ReadResult(register_words)
```
yeah, I might need to use `side_effect` to throw an exception if we decide to get rid of the read cache.

yury-sannikov commented on tests/components/modbus/conftest.py:
```diff
@@ -58,22 +61,23 @@ async def base_test(
     ), mock.patch(
         "homeassistant.components.modbus.modbus.ModbusUdpClient", return_value=mock_sync
+    ), mock.patch(
+        "homeassistant.components.modbus.modbus._read_cached", return_value=read_result
```
hm, not sure why you need to patch pymodbus to make it work. _read_cached basically wraps all read_* calls, that's why I removed the code with the ReadResult below.

yury-sannikov commented:
@janiversen thank you for the review. I will make changes a bit later this week. The main takeaways are:
- remove _read_cached in favor of some sort of read manager handling the sequential reads
- keep non bit_sensor/bit_switch read logic as-is
- refactor unit tests to roll back _read_cached changes
- refactor unit tests to split state check, service call, and asserts on the Modbus write_ calls.
- create separate bit_sensor/bit_switch.py files
lmk if I missed anything.

janiversen commented:
I think those are the big thing, once that is dine I will look at it all again

Your idea of having a base sensor class is very good, please keep that.

yury-sannikov commented:
Hi @janiversen. I think I addressed all the issues except extracting bit_sensor and bit_switch into their own files. That will require extracting the base classes into their own separate files which may increase the complexity of understanding the code.
Also, I separated the read cache from the others. Only bit sensor/switch uses it right now.

janiversen commented:
ok, I will take a look, but it might be a couple of days until I get some free time

janiversen commented:
The modbus integration have changed on dev, so you need to do:
1) git pull on the dev branch
2) git rebase dev on your pr branch
3) git push -f on your pr branch
once done I will do a review

yury-sannikov commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -299,10 +359,12 @@ def _update(self):
                 )
         except ConnectionException:
             self._available = False
+            self.schedule_update_ha_state()
```
w/o this, the state would be STATE_UNKNOWN instead of STATE_UNAVAILABLE. Added tests as well.

janiversen replied:
Please be aware your PR is about bit_sensor, so please only add changes to that respect.

yury-sannikov commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -340,3 +402,72 @@ def _update(self):
+    ):
+        """Initialize the modbus bit sensor."""
+        super().__init__(
+            ModbusReadCache(hub),
```
Only bit switch/sensor uses read cache as a wrapper. Regular sensors/switches would not use read cache.

janiversen replied:
OK. Before you start doing changes, please make sure your branch is rebased against dev, so that this PR only contain your changes (also no merge commits), in order to facilitate review.

Regarding the cache, we are currently thinking about a different more general way. You instanciate an object of each  bit_sensor configured, but an alternative it is modbus.py to have 2 classes one with the real read and one with a virtual read (reusing the first read). That way you would instanciate the first bit_sensor with the real class (reading modbus) and the following with the second class. The reason we are thinking of this, is that we often see a configuration that read address, +1, +2 ..... and that like your bit_sensor is overload.

janiversen commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -37,6 +38,8 @@
     CALL_TYPE_REGISTER_HOLDING,
     CALL_TYPE_REGISTER_INPUT,
+    CONF_BIT_NUMBER,
+    CONF_BIT_SENSORS,
```
This seems to belong in bit_sensor.py

janiversen commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -141,7 +145,34 @@ async def async_setup_platform(
         config = None
 
-    for entry in discovery_info[CONF_SENSORS]:
+    for entry in discovery_info.get(CONF_BIT_SENSORS, []):
```
Please do not mix sensor with bit_sensor !

yury-sannikov replied:
Please see my comment https://github.com/home-assistant/core/pull/47240/files#r604390837

janiversen commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -201,8 +232,8 @@ async def async_setup_platform(
 
-class ModbusRegisterSensor(RestoreEntity, SensorEntity):
-    """Modbus register sensor."""
+class ModbusSensorBase(RestoreEntity, SensorEntity):
```
Are you sure that we should not have 2 base on for register and one for coil ?

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -30,21 +30,25 @@
     MODBUS_DOMAIN,
 )
 from .modbus import ModbusHub
+from .modbus_read_cache import ModbusReadCache
```
This do not belong in switch !

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -184,6 +194,49 @@ def available(self) -> bool:
         """Return True if entity is available."""
         return self._available
 
+    def _read_modbus(self, address) -> int | None:
```
It is very hard to do a review when you move code around (at least that is my impression), please try to make your changes by touching as little as possible in the original code.

When you move code around, we need to look carefully at the moved code to see if it have changed, also later the maintainer will look at the old (but moved code) and see it as new code...meaning they might want changes according to the standards used today.

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -195,49 +248,21 @@ def __init__(self, hub: ModbusHub, config: dict[str, Any]):
     def turn_off(self, **kwargs):
         """Set switch off."""
-        self._write_coil(self._coil, False)
+        self._write_modbus(self._coil, False)
```
This is not correct...there are not write_modbus only write_register and write_coil

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -195,49 +248,21 @@ def __init__(self, hub: ModbusHub, config: dict[str, Any]):
         """Update the state of the switch."""
-        self._is_on = self._read_coil(self._coil)
+        self._is_on = self._read_modbus(self._coil)
         self.schedule_update_ha_state()
```
Same here

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -252,15 +277,13 @@ def __init__(self, hub: ModbusHub, config: dict[str, Any]):
         self._state_off = config.get(CONF_STATE_OFF, self._command_off)
         self._verify_state = config[CONF_VERIFY_STATE]
         self._verify_register = config.get(CONF_VERIFY_REGISTER, self._register)
-        self._register_type = config[CONF_INPUT_TYPE]
```
We need the input type to do the correct reading.

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -299,34 +317,66 @@ def _update(self):
-        self._available = True
+    def __init__(self, hub: ModbusHub, config: dict[str, Any]):
+        """Initialize the register switch."""
+        super().__init__(ModbusReadCache(hub), config)
```
This belong in a separate file.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -44,6 +45,7 @@ async def base_test(
     check_config_only=False,
     config_modbus=None,
     scan_interval=None,
+    mock_hook=None,
```
The hock is something local to conftest.py, we do not want the single test files to care about hooks.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -67,8 +70,11 @@ async def base_test(
+        read_result = (
+            ReadResult(register_words)
+            if register_words
+            else ModbusException("Emulate Modbus read exception")
```
That is a change I really like. That means adding exception testing to all test files.

janiversen commented on tests/components/modbus/test_modbus.py:
```diff
@@ -21,6 +22,19 @@
+    assert number_helper(5.1) == 5.1
+    assert number_helper("5") == 5
+    assert number_helper("5.1") == 5.1
+
```
This does not seem to have anything to do with bit_sensor/bit_switch.

yury-sannikov replied:
just a codecov complain

janiversen replied:
That sounds a bit unlikely. Codecoverage does not complain in branch dev, apart from the fact that this test is already active in test_init.py

yury-sannikov replied:
Removed

janiversen commented on tests/components/modbus/test_modbus_bit_sensor.py:
```diff
@@ -0,0 +1,127 @@
+    sensor_name = "test_sensor"
+    config_sensor = {CONF_NAME: sensor_name, CONF_ADDRESS: 51, CONF_BIT_NUMBER: 5}
+    if do_options:
+        config_sensor.update(
```
Please see in the other test files. if/else in the test cases are to be avoided

yury-sannikov replied:
looks like it's common practice
https://github.com/home-assistant/core/blob/dev/tests/components/modbus/test_modbus_sensor.py#L46
https://github.com/home-assistant/core/blob/dev/tests/components/modbus/test_modbus_switch.py#L62

janiversen replied:
That is old standard, as I noted somewhere else...when you present new code, maintainers look with the present standards. I was asked to remove specifically the “if do_options” in my last PR (and it do make sense). This is a good example why not to move old code.

When I get time, I will change all tests to follow this new pattern.

yury-sannikov replied:
Oh, cool. Sorry, that's my first PR to the HA, I might not know all the rules

janiversen commented on tests/components/modbus/test_modbus_bit_switch.py:
```diff
@@ -0,0 +1,183 @@
+    array_type = None
+    device_config[CONF_ADDRESS] = 1234
+
+    if do_options:
```
Same here

janiversen commented on tests/components/modbus/test_modbus_bit_switch.py:
```diff
@@ -0,0 +1,183 @@
+    """Run test for given config."""
+    switch_name = "modbus_test_switch"
+
+    async def _mock_hook(mock):
```
Having a hook in a specific test case is not a good option. A lot of the platforms allow writing to it needs to be something general

yury-sannikov replied:
I'll probably end up removing those tests. Existing tests do not bother testing state transition

janiversen commented on tests/components/modbus/test_modbus_read_cache.py:
```diff
@@ -0,0 +1,91 @@
+"""Test modbus read cache."""
```
This should be part of the bit_sensor / bit_switch test, we do not want to caching generally.

janiversen reviewed (changes requested):
You have a ton of good code, but there are 2 general issues:
- it seems (maybe wrongly) that you have moved code around, that is tricky
- a number of your changes are not related to bit_sensor/bit_switch
Avoiding those 2 would make review a lot easier and you would have approval a lot faster.

janiversen commented on tests/components/modbus/test_modbus.py:
```diff
@@ -21,6 +22,19 @@
 from .conftest import base_config_test
 
 
+def test_number_helper():
```
Why are you testing number (which is located in __init__.py) here ?

if you look in test_init.py you will see it is tested there.

yury-sannikov commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,9 +184,25 @@ def number(value: Any) -> Union[int, float]:
         vol.Optional(CONF_STATE_OFF): cv.positive_int,
         vol.Optional(CONF_STATE_ON): cv.positive_int,
         vol.Optional(CONF_VERIFY_REGISTER): cv.positive_int,
+        vol.Optional(CONF_VERIFY_STATE, default=True): cv.boolean,
```
Yep, I will remove it

yury-sannikov commented on homeassistant/components/modbus/modbus_read_cache.py:
```diff
@@ -0,0 +1,26 @@
+"""Read cache wrapper for the Modbus Hub methods."""
```
👍🏼

nagyrobi commented:
Will this affect the coils as switches in any aspect?
Any improvement can be done perhaps to read the states in one go?
As far as I can see, on my 24-relay card, each coil state is being read one by one. Whilst with the standard, all of them could be read in one go.

janiversen commented:
@nagyrobi sounds likeyour problem is more what I described in issue #48497 which are in development

yury-sannikov commented:
hi @janiversen 
I created a pretty simple PR out of the `dev` branch which shows the issue I have with the bit_sensor. My goal is to make tests to work against the new bit_sensor. What I see is that `async_setup_platform` is never called.
Could you please take a quick look at the https://github.com/yury-sannikov/core/pull/2/files PR. You might spot the issue I'm fighting for a long time. Thank you!

janiversen commented:
you are basically missing to update modbus.py (the loop that calls all platforms under modbus). I have added some review comments.

yury-sannikov commented:
Thanks for pointing out it. I added suggested changes to the test PR and I'm getting the error I had before: 
```
ERROR:homeassistant.setup:Setup failed for bit_sensor: Integration not found.
```
This happens when I add ` (CONF_BIT_SENSOR, CONF_BIT_SENSORS),` to the list for the `load_platform`

If I do `(CONF_BINARY_SENSOR, CONF_BIT_SENSORS)` instead, binary_sendor integration initialized instead of bit_sensor.

That's why I think there are only 2 ways to hook into:
- add bit_sensor code into binary_senor async_setup_platform fn
- set up a bit_sensor integration (which I feel is an overkill)

I'm looking for the third approach, which I'm struggling to find out.

janiversen commented:
have a look at pr #48558 a fan is also no standard.

yury-sannikov commented:
@janiversen, unfortunately, a fan is a standard domain https://github.com/home-assistant/core/blob/dev/homeassistant/components/fan/manifest.json

yury-sannikov commented:
Any suggestions on how can I go ahead with this PR for the non-standard domain (senor and switch)?

janiversen commented:
For a non-standard domain you need to ask in the developer forum or use discord. But I still do not see why you continue trying to make a non-standard domain.

What you try to do looks like a specialised version of binary_sensor and switch (as I have written earlier). If you cannot trix HA to call you directly (load_component) then 1 single "if" in setup() of binary_sensor.py and switch.py is all that are needed to call you setup and do what you like (well and of course import a function from your file(s)), no need to make things more complicated.

One of the problems I have had reviewing your PR is all your changes in the existing code, causing me to wonder why they are need in order to add a specialised binary_sensor/switch. A clear implementation without all the other stuff is easier to review and easier to guide you on.

Apart from that, your changes in existing code are in conflict with dev, please rebase without merge.

janiversen commented:
Specific example:
```txt
from <yourfile> import <your setup func>

def async_setup_platform(
...
   for entry in ......
       if <your_key> in entry
           <your setup func>
          return
```

you can do it this simple, and without touching other parts of the existing code.

yury-sannikov commented:
Oh, cool, that makes sense now. I thought I'm not allowed to do *any* changes in the binary_sensor.py and switch.py. If I can hook into it like https://github.com/home-assistant/core/pull/47240#issuecomment-813285482 I think I'm good now. Thank you!

janiversen commented:
are you aware that you have changes in 12 files? that sound like overkill for what you want to implement.

may I suggest you do “git rebase” and squash the commit so you have only a few where you easier check which files are changed.

When your PR get merged the maintainer will squash all commits into one, but keep the commit message from every commit.

yury-sannikov commented:
Squashed into one commit. I would appreciate it if you can guide me on how can I reduce the number of changed files. I might be missing something.

janiversen commented:
If you say it is all needed changes, then you cannot reduce the number and at least you have removed 2 files. It looks to as if your changes to conftest.py does not really matter.

I see you are still using "sensor" as base, but as I understand it bit_sensor can return true/false, that is the definition of binary_sensor. I wonder if you will get problems with Lovelace when forcing a sensor (not binary_sensor) to only be true/false.

I see you have introduced your own setup which is good, but I think your code have an unintended sideeffect. For a bit_sensor your setup is called to add the sensor, but then the sensor code continues with the same discovery_info, meaning it is added a second time. (this was just a quick look, I might have overlooked something).

yury-sannikov commented:
> I see you are still using "sensor" as base, but as I understand it bit_sensor can return true/false, that is the definition of binary_sensor. I wonder if you will get problems with Lovelace when forcing a sensor (not binary_sensor) to only be true/false.

Right, forgot about you mentioned it earlier. Let me fix it a bit later.
 
> I see you have introduced your own setup which is good, but I think your code have an unintended sideeffect. For a bit_sensor your setup is called to add the sensor, but then the sensor code continues with the same discovery_info, meaning it is added a second time. (this was just a quick look, I might have overlooked something).

`bit_sensor` uses `discovery_info.get(CONF_BIT_SENSORS, [])` and `sensor` use `discovery_info.get(CONF_SENSORS, [])` Should be no dupes even if both `setup_bit_sensors` and the code after it adds the sensors

yury-sannikov commented:
Hi @janiversen
I believe I addressed all the issues. I checked the code against my hardware setup and it's working as expected. If you have some time, could you pls do another round of review?

janiversen commented:
2 general items:
- please update the description the PR you refer is merged.
- please update the doc. PR as frenck requested a while ago.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+    """Set up the Modbus Bit sensors."""
+    sensors = []
+
+    if discovery_info is None:
```
This check really belongs in binary_sensor.py.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+
+def setup_bit_sensors(
+    hass: HomeAssistantType,
+    discovery_info: DiscoveryInfoType | None = None,
```
Why this complicated definition, it is called from binary_sensor, so you know exactly how it is called.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+    if discovery_info is None:
+        return sensors
+
+    for entry in discovery_info.get(CONF_BIT_SENSORS, []):
```
Why use info.get that seems like an overkill (setup should only be called if BIT_SENSOR is present).

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+        words_count = int(entry[CONF_COUNT])
+        bit_number = int(entry[CONF_BIT_NUMBER])
+
+        if bit_number >= words_count * 16:
```
Seems the 16 are related to the 32 you uses earlier, would be nice to have this as a instead, especially since they seem related.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+            )
+            continue
+
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
```
CONF_NAME is an optional parameter !

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+            ModbusBitSensor(
+                hub,
+                entry[CONF_NAME],
+                entry.get(CONF_SLAVE),
```
No default ? The scheme has a default.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+                entry[CONF_INPUT_TYPE],
+                bit_number,
+                words_count,
+                entry.get(CONF_DEVICE_CLASS),
```
No default ?

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+    return sensors
+
+
+class ModbusBitSensorBase(RestoreEntity, BinarySensorEntity):
```
Why do you need a base ? Seems like you could merge ModbusBitSensor and MobusBitSensorBase.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+        device_class,
+        scan_interval,
+    ):
+        """Initialize the modbus sensor."""
```
Bit sensor ?

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+
+
+class ModbusBitSensorBase(RestoreEntity, BinarySensorEntity):
+    """Base class for the Modbus sensor."""
```
Bit sensor ?

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+    @property
+    def is_on(self):
+        """Return the state of the sensor."""
+        return self._value
```
What happens if _value is None, HA expects true/false.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+        """Update the state of the sensor."""
+        try:
+            if self._register_type == CALL_TYPE_REGISTER_INPUT:
+                result = self._hub.read_input_registers(
```
This seems like one of the important lines to test ??

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+                    self._slave, self._register, self._count
+                )
+
+        except ConnectionException:
```
Just check for ModbusException.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+            self.schedule_update_ha_state()
+            return
+
+        register_index = self._bit_number // 16
```
Do not use 16 use a constant.

janiversen commented on homeassistant/components/modbus/bit_sensor.py:
```diff
@@ -0,0 +1,223 @@
+            return
+
+        register_index = self._bit_number // 16
+        register_bit_mask = 1 << (self._bit_number % 16)
```
Same here.

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+import logging
+from typing import Any
+
+from pymodbus.exceptions import ConnectionException, ModbusException
```
You only need ModbusException (which is also on its way out, pending merge of my PR).

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+from typing import Any
+
+from pymodbus.exceptions import ConnectionException, ModbusException
+from pymodbus.pdu import ExceptionResponse
```
Same here.

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+    """Modbus Bit Switches setup."""
+    switches = []
+
+    if discovery_info is None:
```
This function should not be called if discovery_info is None (you are the only one calling it).

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+    if discovery_info is None:
+        return switches
+
+    for entry in discovery_info.get(CONF_BIT_SWITCHES, []):
```
Setup should only be called if there are BIT_SWITCHES so no need for info.get()

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+        return switches
+
+    for entry in discovery_info.get(CONF_BIT_SWITCHES, []):
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
```
Remark CONF_NAME is optional.

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+    return switches
+
+
+class ModbusBaseSwitch(SwitchEntity, RestoreEntity, ABC):
```
ModbusBaseSwitch (if needed) do not belong in this file !

If it is something only used for bit_switch, you should either use ModbusBaseSwitch from switch.py or merge this class with modbusBit_switch.

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+            else:
+                result = self._hub.read_holding_registers(self._slave, address, 1)
+
+        except ConnectionException:
```
No need for this check.

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+            self._is_on = bool(value & self._status_bit_mask)
+            self._available = True
+        else:
+            self._available = False
```
But this is already done in _read...

Setting of _available seems duplicated in many places and inconsistent.

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+            return
+        register_value = self._read_register(self._verify_register)
+        if register_value is None:
+            self._available = False
```
But this is already done in _read_...

janiversen commented on homeassistant/components/modbus/bit_switch.py:
```diff
@@ -0,0 +1,201 @@
+            return
+        register_value = self._read_register(self._verify_register)
+        if register_value is None:
+            self._available = False
```
Same here.

janiversen commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -125,7 +125,7 @@ async def async_setup_platform(
     sensors = []
 
-    #  check for old config:
+    # check for old config:
```
Does this change have anything to do with bit_...

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -93,7 +94,7 @@ async def async_setup_platform(
     switches = []
 
-    #  check for old config:
+    # check for old config:
```
Does this change have anything to do with bit_....

janiversen commented on homeassistant/components/modbus/switch.py:
```diff
@@ -121,7 +122,9 @@ async def async_setup_platform(
-    for entry in discovery_info[CONF_SWITCHES]:
+    switches.extend(setup_bit_swithes(hass, discovery_info))
+
+    for entry in discovery_info.get(CONF_SWITCHES, []):
```
No need to change this to info.get() old way worked nicely.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -38,7 +38,7 @@ async def base_test(
     array_name_discovery,
     array_name_old_config,
-    register_words,
+    register_words_or_exception,
```
Please do not change the test methods in a PR where you implement new functionality. In case of problems it is quite hard to determine whether it is the new functionaliy or the test changes.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -58,6 +58,7 @@ async def base_test(
         }
 
     mock_sync = mock.MagicMock()
+    # Setup inputs for the sensor
```
?? This is not just for a sensor.

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -67,8 +68,11 @@ async def base_test(
+            register_words_or_exception
+            if isinstance(register_words_or_exception, Exception)
+            else ReadResult(register_words_or_exception)
+        )
```
This is a good idea, but we do not want to test pymodbus in every platform, that is the responsibiliy and thus testing of modbus.py.

yury-sannikov replied:
It's not a Modbus test, it's a negative case testing wish Modbus integration lack of, unfortuately

janiversen commented on tests/components/modbus/conftest.py:
```diff
@@ -118,7 +122,6 @@ async def base_test(
             async_fire_time_changed(hass, now)
             await hass.async_block_till_done()
 
-        # Check state
```
Why remove this comment ?

janiversen commented on tests/components/modbus/test_modbus_bit_sensor.py:
```diff
@@ -0,0 +1,337 @@
+        ),
+    ],
+)
+@mock.patch.object(ModbusHub, "read_holding_registers")
```
There are a standard function in conftest you should use.

janiversen reviewed (changes requested):
I have not looked too closely at your test cases, it seems to me you are redefining the test harness we use for other platforms, and that makes the code a lot harder to read.

yury-sannikov commented:
Most of the review comments do not make much sense to me. Not sure I will ever be able to make it to conform to the controversial suggestions here. Feel free to use some portions of the code for further development of the platform. I will focus my energy on more meaningful things. Thanks for your time.
